dev/triggers.h

```diff
@@ -0,0 +1,204 @@
+#pragma once
+
+#include <memory>
+#include <sstream>
+#include <string>
+#include <tuple>
+
+#include "tuple_helper/tuple_helper.h"
+
+// NOTE Idea : Maybe also implement a custom trigger system to call a c++ callback when a trigger triggers ?
+// (Could be implemented with a normal trigger that insert or update an internal table and then retreive
+// the event in the C++ code, to call the C++ user callback, with update hooks: https://www.sqlite.org/c3ref/update_hook.html)
+// It could be an interesting feature to bring to sqlite_orm that other libraries don't have ?
+
+namespace sqlite_orm {
+    namespace internal {
+        enum class trigger_timing { trigger_before, trigger_after, trigger_instead_of };
+        enum class trigger_type { trigger_delete, trigger_insert, trigger_update };
+
+        /**
+         * This class is an intermediate SQLite trigger, to be used with
+         * `make_trigger` to create a full trigger.
+         *  T is the base of the trigger (contains its type, timing and associated table)
+         *  S is the list of trigger statments
+         */
+        template<class T, class... S>
+        struct partial_trigger_t {
+            using statements_type = std::tuple<S...>;
+
+            /**
+             * Statements of the triggers (to be executed when the trigger fires)
+             */
+            statements_type statements;
+            /**
+             * Base of the trigger (contains its type, timing and associated table)
+             */
+            T base;
+
+            partial_trigger_t(T trigger_base, S... statements) :
+                statements(std::make_tuple<S...>(std::forward<S>(statements)...)), base(trigger_base) {}
+
+            partial_trigger_t &end() {
+                return *this;
+            }
+        };
+
+        /**
+         * This class represent a SQLite trigger
+         *  T is the base of the trigger (contains its type, timing and associated table)
+         *  S is the list of trigger statments
+         */
+        template<class T, class... S>
+        struct trigger_t {
+            using object_type = void;  // Not sure
+            using elements_type = typename partial_trigger_t<T, S...>::statements_type;
+
+            /**
+             * Name of the trigger
+             */
+            std::string name;
+            /**
+             * Contains more data about the trigger
+             */
+            partial_trigger_t<T, S...> part;
+            /**
+             * Statements of the triggers (to be executed when the trigger fires)
+             * (reference to partial_trigger::statements for conveniance)
+             */
+            const elements_type &elements;
```

Comment:
I put the reference here, because in storage_impl's foreign_keys_count implementation, the table is required to have a 'elements' field (not a function) and I didn't want to copy it.

Should I change `const elements_type &elements` to just `elements_type elements` instead ?